### ISSUE TITLE
recursion issue

### ISSUE THREAD

**Issue by sabmah** [agileobjects/AgileMapper#62]
Converting From

```
public class Location
{
  public int Id { get; set; } 
  public Place Place { get; set; }
}
public class Place
{
  public IEnumerable<Location> Locations { get; set; }
}
```

To my Dto model 

```
public class Location
{
  public int Id { get; set; } 
  public Place Place { get; set; }
}
public class Place
{
  public IEnumerable<Location> Locations { get; set; }
}
```
This is my plan on startup

```
AgileObjects.AgileMapper.Mapper
                .GetPlanFor<Location>()
                .ToANew<Dto.Location>();

            AgileObjects.AgileMapper.Mapper
                .GetPlanFor<Place>()
                .ToANew<Dto.Place>();


```

```
AgileObjects.AgileMapper.Mapper.Map<IEnumerable<Location>>(data).ToANew<IEnumerable<Dto.Location>>();
```

I have following error

```
<Error>
<Message>An error has occurred.</Message>
<ExceptionMessage>
An exception occurred mapping IEnumerable<Location> -> IEnumerable<Location> with rule set CreateNew.
</ExceptionMessage>
<ExceptionType>AgileObjects.AgileMapper.MappingException</ExceptionType>
<StackTrace>

</StackTrace>
<InnerException>
<Message>An error has occurred.</Message>
<ExceptionMessage>
An exception occurred mapping IEnumerable<Location>[i] -> IEnumerable<Location>[i] with rule set CreateNew.
</ExceptionMessage>
<ExceptionType>AgileObjects.AgileMapper.MappingException</ExceptionType>
<StackTrace>at lambda_method(Closure , IObjectMappingData`2 )</StackTrace>
<InnerException>
<Message>An error has occurred.</Message>
<ExceptionMessage>
An exception occurred mapping IEnumerable<Location>[i].Place -> IEnumerable<Location>[i].Place with rule set CreateNew.
</ExceptionMessage>
<ExceptionType>AgileObjects.AgileMapper.MappingException</ExceptionType>
<StackTrace>at lambda_method(Closure , IObjectMappingData`2 )</StackTrace>
<InnerException>
<Message>An error has occurred.</Message>
<ExceptionMessage>
An exception occurred mapping IEnumerable<Location>.Place.Locations -> IEnumerable<Location>[i].Place.Locations with rule set CreateNew.
</ExceptionMessage>
<ExceptionType>AgileObjects.AgileMapper.MappingException</ExceptionType>
<StackTrace>
at lambda_method(Closure , IObjectMappingData`2 ) at AgileObjects.AgileMapper.ObjectPopulation.Recursion.RecursionMapperFunc`2.Map(IObjectMappingData mappingData) at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapper`2.MapRecursion(IObjectMappingData childMappingData) at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.AgileObjects.AgileMapper.ObjectPopulation.IObjectMappingDataUntyped.MapRecursion[TDeclaredSource,TDeclaredTarget](TDeclaredSource sourceValue, TDeclaredTarget targetValue, Nullable`1 enumerableIndex, String targetMemberName, Int32 dataSourceIndex) at lambda_method(Closure , IObjectMappingData`2 )
</StackTrace>
<InnerException>
<Message>An error has occurred.</Message>
<ExceptionMessage>
An exception occurred mapping IEnumerable<Location>.Place.Locations[i] -> IEnumerable<Location>[i].Place.Locations[i] with rule set CreateNew.
</ExceptionMessage>
<ExceptionType>AgileObjects.AgileMapper.MappingException</ExceptionType>
<StackTrace>at lambda_method(Closure , IObjectMappingData`2 )</StackTrace>
<InnerException>
<Message>An error has occurred.</Message>
<ExceptionMessage>
Object reference not set to an instance of an object.
</ExceptionMessage>
<ExceptionType>System.NullReferenceException</ExceptionType>
<StackTrace>
at AgileObjects.AgileMapper.Caching.ArrayCache`2.GetOrAdd(TKey key, Func`2 valueFactory) at AgileObjects.AgileMapper.ObjectPopulation.ObjectMapper`2.MapRecursion(IObjectMappingData childMappingData) at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.AgileObjects.AgileMapper.ObjectPopulation.IObjectMappingDataUntyped.MapRecursion[TDeclaredSource,TDeclaredTarget](TDeclaredSource sourceValue, TDeclaredTarget targetValue, Nullable`1 enumerableIndex, String targetMemberName, Int32 dataSourceIndex) at AgileObjects.AgileMapper.ObjectPopulation.ObjectMappingData`2.AgileObjects.AgileMapper.ObjectPopulation.IObjectMappingDataUntyped.MapRecursion[TDeclaredSource,TDeclaredTarget](TDeclaredSource sourceValue, TDeclaredTarget targetValue, Nullable`1 enumerableIndex, String targetMemberName, Int32 dataSourceIndex) at lambda_method(Closure , IObjectMappingData`2 )
</StackTrace>
</InnerException>
</InnerException>
</InnerException>
</InnerException>
</InnerException>
</Error>
```

**Comment by SteveWilkes** [comment]
Hi,

Thanks very much for the feedback - there is a bug here which I'll fix - but in the meantime, your mapping will work if you do it like this:

```C#
// Cache the plan for mapping an IEnumerable<Location> to an IEnumerable<Dto.Location>;
// Plans for mapping a Location to a Dto.Location and a Place to a Dto.Place will automatically
// be generated as part of this:
AgileObjects.AgileMapper.Mapper
    .GetPlanFor<IEnumerable<Location>>()
    .ToANew<IEnumerable<Dto.Location>>();
```

I've written [a .NET Fiddle](https://dotnetfiddle.net/BD34QO) which caches and performs the mapping you're doing as an example.

**Comment by sabmah** [comment]
@SteveWilkes Thank for the reply. I got around it by doing this:

```
Locations.Select(x =>AgileObjects.AgileMapper.Mapper.Map(x.Place).ToANew<Dto.Location>)
```

**Comment by SteveWilkes** [comment]
Fixed in latest code, release to follow.

**Comment by SteveWilkes** [comment]
Fixed in [v0.23.4](https://www.nuget.org/packages/AgileObjects.AgileMapper). Thanks again for the feedback! :)

**Comment by sabmah** [comment]
@SteveWilkes  Thanks for the update :)
